x/gov/handler_test.go

```diff
@@ -581,35 +592,264 @@ func TestTally(t *testing.T) {
+
+	specs := map[string]struct {
+		Msg            UpdateElectorateMsg
+		SignedBy       weave.Condition
+		WantCheckErr   *errors.Error
+		WantDeliverErr *errors.Error
+		ExpModel       *Electorate
+		Mods           ctxAwareMutator // modifies TextProposal test fixtures before storing
+	}{
+		"All good with update by owner": {
+			Msg: UpdateElectorateMsg{
+				ElectorateID: electorateID,
+				Electors:     []Elector{{Address: alice, Weight: 22}},
+			},
+			SignedBy: bobbyCond,
+			ExpModel: &Electorate{
+				Admin:                 bobby,
+				Title:                 "fooo",
+				Electors:              []Elector{{Address: alice, Weight: 22}},
+				TotalWeightElectorate: 22,
+			},
+		},
+		"Update by non owner should fail": {
+			Msg: UpdateElectorateMsg{
+				ElectorateID: electorateID,
+				Electors:     []Elector{{Address: alice, Weight: 22}},
+			},
+			SignedBy:       aliceCond,
+			WantCheckErr:   errors.ErrUnauthorized,
+			WantDeliverErr: errors.ErrUnauthorized,
+		},
+		"Update with open proposal should fail": {
+			Msg: UpdateElectorateMsg{
+				ElectorateID: electorateID,
+				Electors:     []Elector{{Address: alice, Weight: 22}},
+			},
+			SignedBy: bobbyCond,
+			ExpModel: &Electorate{
+				Admin:                 bobby,
+				Title:                 "fooo",
+				Electors:              []Elector{{Address: alice, Weight: 22}},
+				TotalWeightElectorate: 22,
+			},
+			Mods: func(ctx weave.Context, proposal *TextProposal) {
+				proposal.Status = TextProposal_Submitted
+			},
+			WantDeliverErr: errors.ErrInvalidState,
+		},
+		"Update with too many electors should fail": {
+			Msg: UpdateElectorateMsg{
+				ElectorateID: electorateID,
+				Electors:     buildElectors(2001),
+			},
+			SignedBy:       bobbyCond,
+			WantCheckErr:   errors.ErrInvalidInput,
+			WantDeliverErr: errors.ErrInvalidInput,
+		},
+		"Update without electors should fail": {
+			Msg: UpdateElectorateMsg{
+				ElectorateID: electorateID,
+			},
+			SignedBy:       bobbyCond,
+			WantCheckErr:   errors.ErrEmpty,
+			WantDeliverErr: errors.ErrEmpty,
+		},
+		"Duplicate electors should fail": {
+			Msg: UpdateElectorateMsg{
+				ElectorateID: electorateID,
+				Electors:     []Elector{{Address: alice, Weight: 1}, {Address: alice, Weight: 2}},
+			},
+			SignedBy:       bobbyCond,
+			WantCheckErr:   errors.ErrInvalidInput,
+			WantDeliverErr: errors.ErrInvalidInput,
+		},
+		"Empty address in electors should fail": {
+			Msg: UpdateElectorateMsg{
+				ElectorateID: electorateID,
+				Electors:     []Elector{{Address: weave.Address{}, Weight: 1}},
+			},
+			SignedBy:       bobbyCond,
+			WantCheckErr:   errors.ErrEmpty,
+			WantDeliverErr: errors.ErrEmpty,
+		},
+	}
+	bucket := NewElectorateBucket()
+	for msg, spec := range specs {
+		t.Run(msg, func(t *testing.T) {
+			auth := &weavetest.Auth{
+				Signer: spec.SignedBy,
+			}
+			rt := app.NewRouter()
+			RegisterRoutes(rt, auth)
+			db := store.MemStore()
+			withElectorate(t, db)
+			if spec.Mods != nil {
```

Comment:
Ahh... this is the point, you only create the electorate if there is a mod.
It would be clearer to use a boolean "with proposal" or something. Or rename: CreateWithMods.

As it, not too clear.

Since you have Mods, can you also add the case where there is one proposal that is closed already (voting finished) and validate that electorate *can* be changed in that case?

Reply:
@ethanfrey fixed. Yeah, it is indeed a bit confusing